app/controllers/spree/orders_controller.rb

```diff
@@ -69,6 +69,10 @@ def update
       if @order.contents.update_cart(order_params)
         @order.recreate_all_fees! # Enterprise fees on line items and on the order itself
 
+        # Re apply the voucher
+        VoucherAdjustmentsService.calculate(@order)
+        @order.update_totals_and_states
```

Comment:
Shouldn't this happen after updating the payment fees? Ah, no, this can reduce the payment. Argh, so, do we exclude payment fees from the voucher calculation?

Reply:
So if I understand correctly, 
* the payment fees will be applied when we move to the confirmation step 
* voucher calculation happen after transitioning to the confirmation step (via state machine `after_transition` hook) -> so payment fees are included in the voucher calculation

The code above is used when updating the order from the `/cart` page, and as far as I can see updating anything reset the checkout flow to "details". So we end up in the scenario above and payment fee will be included in the voucher calculation.

Reply:
Okay, so let's play this through. I have an order for $100. The Stripe payment adds a 2% fee: $2. So the total then is $102.

Then the a 10% voucher is applied, value would be $10.20. And the amount to pay reduces to $91.80.

I think that the checkout creates or updates a payment with its fee. That reduces the payment fee from $2 to $1.836 -> $1.84 and it updates the order total by 16 cents to $91.64. That amount should get charged. We have an unfortunate interdependence between voucher value and payment fee which can't be resolved completely.

Once an order is complete, there's a before_save callback to update payment fees and the same is happening on the cart page. It makes it more confusing but we don't really need to look at that now, i think.

In summary, I think that we are doing the best we can. Oh, wait, a new thought: we need to calculate the payment fee to be displayed to the user but the payment fee is really based on the payment which happens last. They were only introduced because of the fees by Stripe and Paypal. So if we exclude payment fees from the voucher calculation and calculate the payment fee after the voucher was applied then we should have the correct amount, right?

Reply:
:exploding_head: I'll have another look

Reply:
So I dig into that, to the extend of my knowledge payment fees are calculated based on total of items, excluding any other fees. So even in the case of a voucher being applied, the payment fee won't change when it's recalculated once the order is completed.
To take an example : 
With Order $100 (we exclude any other fees for the sake of simplicity), a Voucher of 10% and a Payment fee of 10%
- We apply the payment fee => 100 + (100 * 0.1) = 110
- We apply the voucher  => 110 - (110 * 0.1) = 99
=> Order break down : $100 product + $10 payment fees - $11 voucher = $99

Once the order is completed, the payment fees are recalculated, be because the voucher doesn't affect the item total it stays the same. So I think we are good ?

Reply:
@lin-d-hop @kirstenalarsen could you confirm if the above is correct ?

Reply:
@rioug this looks right to me. I agree with your findings of fees calculated on line items and thus not creating the circular behaviour that was feared.
The key here will be for rigourous @openfoodfoundation/testers ;)
We need to test the edge cases where the total is close to zero. We need to test updates to orders throughout the process to make sure all fees and vouchers are updated. Perhaps off the back of this @rioug you could spend a few minutes thinking of edge cases and confirming the 'what to test' section is right :pray:

Reply:
OK, I updated the testing notes to include payment fees, and I also updated the testing notes for #10821 which is really the PR where this discussion should be taking place,

Reply:
Ah, okay, I had the wrong assumption of the payment fee calculation then. I'm glad it's not an issue.